// react-pick/lib/ListPopupOption.js
"use strict";

var _objectWithoutProperties = function (obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; };

var React = require("react/addons");

var PureRenderMixin = React.addons.PureRenderMixin;

/**
 * <ListPopupOption> is a default implementation for rows rendered for the 
 * `options` in the <ListPopup> component. For the row content, it displays the 
 * string label as determined by the `getLabelForOption` function.
 */
var ListPopupOption = React.createClass({
  displayName: "ListPopupOption",

  mixins: [PureRenderMixin],

  propTypes: {
    /**
     * Function that takes an `option` value, and returns a string label.
     */
    getLabelForOption: React.PropTypes.func.isRequired,

    /**
     * Option value that the row is representing.
     */
    option: React.PropTypes.any.isRequired
  },

  render: function render() {
    var _props = this.props;
    var option = _props.option;
    var getLabelForOption = _props.getLabelForOption;

    var otherProps = _objectWithoutProperties(_props, ["option", "getLabelForOption"]);

    return React.createElement(
      "div",
      otherProps,
      getLabelForOption(option)
    );
  }

});

module.exports = ListPopupOption;
